engine/crates/engine-v2/src/plan/attribution.rs

```diff
@@ -129,181 +132,121 @@ impl<'a> std::ops::Deref for ExtraFieldWalker<'a> {
-                                        }
-                                        index += 1;
-                                    }
-                                } else {
-                                    name
+    fn insert_extra_selection_set(&mut self, extra: ExtraBoundarySelectionSet) -> ExtraSelectionSetId {
+        let selection_set = ExtraSelectionSet {
+            ty: extra.ty,
+            fields: extra
+                .fields
+                .into_values()
+                .filter_map(
+                    |ExtraBoundaryField {
+                         extra_field,
+                         read: used,
+                     }| {
+                        if used {
+                            Some(extra_field)
+                        } else {
+                            None
+                        }
+                    },
+                )
+                .map(
+                    |ExtraField {
+                         edge,
+                         type_condition,
+                         field_id,
+                         expected_key,
+                         ty,
+                     }| {
+                        let field = ExtraField {
+                            edge,
+                            type_condition,
+                            field_id,
+                            expected_key,
+                            ty: match ty {
+                                ExpectedType::Scalar(scalar) => ExpectedType::Scalar(scalar),
+                                ExpectedType::SelectionSet(extra) => {
+                                    ExpectedType::SelectionSet(self.insert_extra_selection_set(extra))
                                 }
-                            })
-                            .to_string()
-                    } else {
-                        field.name().to_string()
-                    }
-                },
-                ty: field
-                    .ty()
-                    .inner()
-                    .data_type()
-                    .map(ExpectedType::Scalar)
-                    .unwrap_or_else(|| {
-                        let id = ExtraSelectionSetId::from(self.extra_selection_sets.len());
-                        self.extra_selection_sets.push(ExtraSelectionSetBuilder {
-                            ty: SelectionSetType::maybe_from(field.ty().inner().id()).unwrap(),
-                            fields: HashMap::new(),
-                        });
-                        ExpectedType::SelectionSet(id)
-                    }),
-            });
-            self[extra_selection_set_id].fields.insert(field_id, extra_field_id);
-            extra_field_id
-        } else {
-            self[extra_selection_set_id].fields[&field_id]
+                            },
+                        };
+                        let id = ExtraFieldId::from(self.extra_fields.len());
+                        self.extra_fields.push(field);
```

Comment:
I find this a bit too confusing to mutate `self` in an iterator map. Maybe convert this whole structure to a nice for loop for clarity?